torchtitan/utils.py

```diff
@@ -52,26 +53,75 @@ def _warn_overwrite_env(env, val):
     os.environ[env] = val
 
 
-def set_determinism(seed: Optional[int]) -> None:
+def set_determinism(
+    world_mesh: DeviceMesh,
+    device: torch.device,
+    job_config: JobConfig,
+) -> None:
     """
-    Set Python, PyTorch, CUDA seeds and cudnn settings for reproducibility
+    Set the same DTensor manual seed for all ranks within the same DTensor SPMD group, but different
+    seeds across PP groups (if applicable).
+
+    TODO: Link to documentation for DTensor RNG setup? Or explain more here?
+
+    Currently, does not set seeds for the CUDA RNG since TorchTitan always uses DTensor for SPMD parallelisms,
+    and DTensor manages its own RNG tracker, but we could extend to support both if needed.
+
+    Set Determinism flags for increased reproducibility with loss of performance.
     """
+    deterministic = job_config.training.deterministic
+    seed = job_config.training.seed
+
+    if deterministic:
+        logger.info("Deterministic training enabled (expect perf degradation).")
+        torch.use_deterministic_algorithms(True)
+        # env var for deterministic CuBLAS
+        # https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html
+        os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
+    torch.backends.cudnn.deterministic = deterministic
+    torch.backends.cudnn.benchmark = not deterministic
+
+    # to ensure we can control which ranks have same or different seeds, all ranks agree on a starting seed.
+    # if user provides one, we use this. Otherwise rank 0 rolls the dice and everyone else uses that.
     if seed is not None:
         # CPU and GPU determinism
         torch.manual_seed(seed)
```

Comment:
Could you remind me of the logic here? It seems right now:
1. every rank would call `torch.manual_seed` on the same seed, regardless of PP existence / ranks
2. different PP ranks would call `torch.distributed.tensor._random.manual_seed` on different seeds, on the spmd mesh.

What does `torch.manual_seed` control in torchtitan?